File: Third Semester - Advanced Programming Methods/Lab2/decorator/StrategyTaskRunner.java
package decorator;

import factory.Factory;
import model.Container;
import factory.Strategy;
import factory.TaskContainerFactory;
import model.Task;

public class StrategyTaskRunner implements TaskRunner {
    private final Container container;

    public StrategyTaskRunner(Strategy strategy) {
        Factory factory = TaskContainerFactory.getInstance();
        container = factory.createContainer(strategy);
    }

    @Override
    public void executeOneTask() {
        Task task = container.remove();
        task.execute();
    }

    @Override
    public void executeAll() {
        while (hasTask()) {
            executeOneTask();
        }
    }

    @Override
    public void addTask(Task t) {
        container.add(t);
    }

    @Override
    public boolean hasTask() {
        return !container.isEmpty();
    }
}

File: Third Semester - Advanced Programming Methods/Lab10/decorator/StrategyTaskRunner.cs
using lab10.factory;
using lab10.model;
using Task = lab10.model.Task;
namespace lab10.decorator;

public class StrategyTaskRunner: ITaskRunner
{
    private readonly IContainer _container;

    public StrategyTaskRunner(Strategy strategy)
    {
        IFactory factory = TaskContainerFactory.Instance;
        _container = factory.createContainer(strategy);
    }

    public void ExecuteOneTask()
    {
        Task task = _container.Remove();
        task.execute();
    }

    public void ExecuteAllTasks()
    {
        while (HasTask())
        {
            ExecuteOneTask();
        }
    }

    public void AddTask(Task task)
    {
        _container.Add(task);
    }

    public bool HasTask()
    {
        return !_container.IsEmpty();
    }
}